samples/extensions/sparse_image/sparse_image.cpp

```diff
@@ -0,0 +1,1543 @@
+		virtual_texture.sparse_image_memory_bind[page_index].extent.width  = (virtual_texture.mip_properties[mipLevel].width - virtual_texture.sparse_image_memory_bind[page_index].offset.x < virtual_texture.format_properties.imageGranularity.width) ? virtual_texture.mip_properties[mipLevel].width - virtual_texture.sparse_image_memory_bind[page_index].offset.x : virtual_texture.format_properties.imageGranularity.width;
+		virtual_texture.sparse_image_memory_bind[page_index].extent.height = (virtual_texture.mip_properties[mipLevel].height - virtual_texture.sparse_image_memory_bind[page_index].offset.y < virtual_texture.format_properties.imageGranularity.height) ? virtual_texture.mip_properties[mipLevel].height - virtual_texture.sparse_image_memory_bind[page_index].offset.y : virtual_texture.format_properties.imageGranularity.height;
+	}
+
+	//==================================================================================================
+	// Allocating memory
+
+	VkMemoryAllocateInfo memory_allocate_info{};
+	memory_allocate_info.sType           = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+	memory_allocate_info.allocationSize  = virtual_texture.page_size * virtual_texture.mip_properties[0].mip_num_pages / 2U;        // TODO(GS): what is the minimum memory required
+	memory_allocate_info.memoryTypeIndex = get_device().get_memory_type(virtual_texture.mem_requirements.memoryTypeBits, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+
+	for (size_t i = 0U; i < (virtual_texture.mip_properties[0].mip_num_pages / 2U); i++)        // TODO(GS): what is the minimum memory required
+	{
+		virtual_texture.available_memory_index_list.push_back(i);
+	}
+
+	VK_CHECK(vkAllocateMemory(get_device().get_handle(), &memory_allocate_info, nullptr, &virtual_texture.texture_memory));
+
+	//==================================================================================================
+	// Creating texture image view
+
+	VkImageViewCreateInfo view_info           = vkb::initializers::image_view_create_info();
+	view_info.image                           = virtual_texture.texture_image;
+	view_info.viewType                        = VK_IMAGE_VIEW_TYPE_2D;
+	view_info.format                          = VK_FORMAT_R8G8B8A8_SRGB;
+	view_info.subresourceRange.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
+	view_info.subresourceRange.baseMipLevel   = virtual_texture.base_mip_level;
+	view_info.subresourceRange.levelCount     = virtual_texture.mip_levels;
+	view_info.subresourceRange.baseArrayLayer = 0U;
+	view_info.subresourceRange.layerCount     = 1U;
+
+	VK_CHECK(vkCreateImageView(get_device().get_handle(), &view_info, nullptr, &virtual_texture.texture_image_view));
+
+	for (uint8_t mip_level = 0U; mip_level <= virtual_texture.mip_levels - 1U; mip_level++)
+	{
+		transition_mip_layout(virtual_texture.texture_image, VK_IMAGE_LAYOUT_UNDEFINED, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, mip_level);
```

Comment:
Good point. Actually, this particular piece of code is currently used only because of how the state machine is implemented. That is, each stage is executed once per frame. Execution starts at (let's say) Stage2 whereas a proper data transfers and layout transitions are happening on the (let's say) Stage4. Between those two stages, there are 2-3 frames rendered. Without this particular transition happening, there are validation errors reported because the image is in an `VK_IMAGE_LAYOUT_UNDEFINED` state although being rendered.

This issue is only relevant for the first, complete state-machine cycle and those 2-3 initial frames. I will think about some way to run the whole cycle before the first frame, or/and change the initial image layout. Anyway, yes, it is going to be changed.

Reply:
As I mentioned, I changed the process, so that the least detailed LOD is loaded before the first frame is rendered. Additionally, I changed the initial layout of an image to `VK_IMAGE_LAYOUT_PREINITIALIZED` which should not negatively interfere with a transition to `VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL`.